src/classes/Contacts.cls

```diff
@@ -0,0 +1,691 @@
+            //and send it to our naming code for action
+            //first contact's names to their old names, and see if anything changed
+            //salutation
+            if (!hasHousehold(c)) {
+                handleContactHasNoHousehold(c);
+            }
+
+            //if they've changed households, we need to query the old
+            //household to see if anybody is left, if so, send it for renaming
+            //if not, delete it.
+            if (householdChanged(c)) {
+                handleHouseholdChanged(c);
+            } else if (isNamingExclusionsChanged(c) || isNamingFieldChanged(c)) {
+                householdsNeedingNamesUpdated.add(c.npo02__household__c);
+            }
+        }
+    }
+
+    Set<String> householdNamingFieldsInUse;
+    private Boolean isNamingFieldChanged(Contact c) {
+        if (householdNamingFieldsInUse == null) {
+            householdNamingFieldsInUse =
+                    new HH_HouseholdNaming().getHouseholdNamingContactFields();
+        }
+
+        Boolean namingFieldChanged = false;
+        if (hasHousehold(c)) {
+            for (String fieldName : householdNamingFieldsInUse) {
+                Boolean fieldChanged = isFieldChanged(c, fieldName);
+                if (fieldChanged) {
+                    namingFieldChanged = true;
+                    break;
+                }
+            }
+        }
+        return namingFieldChanged;
+    }
+
+    private Boolean isFieldChanged(Contact c, String fieldName) {
+        // make sure to use case sensitive comparison
+        String strName = getNewFieldValue(c, fieldName);
+        String strNameOld = getOldFieldValue(c, fieldName);
+        Boolean isChanged = !strName.equals(strNameOld);
+        return isChanged;
+    }
+
+    private Boolean hasHousehold(Contact c) {
+        return c.npo02__household__c != null;
+    }
+
+    private Boolean isNamingExclusionsChanged(Contact c) {
+        return c.npo02__Naming_Exclusions__c != oldMap.get(c.id).npo02__Naming_Exclusions__c;
+    }
+
+    private void handleHouseholdChanged(Contact c) {
+        householdsNeedingNamesUpdated.add(c.npo02__household__c);
+        householdIdsNeedingRollup.add(c.npo02__household__c);
+
+        Id oldHouseholdId = getOldHouseholdId(oldMap, c);
+        if (oldHouseholdNeedsRefresh(oldHouseholdSizeById, oldMap, c)) {
+            householdsNeedingNamesUpdated.add(oldHouseholdId);
+            householdIdsNeedingRollup.add(oldHouseholdId);
+        } else if (oldHouseholdIsEmpty(oldHouseholdSizeById, oldMap, c)) {
+            emptyHouseholdsToDelete.add(new npo02__Household__c(Id = oldHouseholdId));
+        }
+    }
+
+    private Boolean oldHouseholdIsEmpty(Map<Id, Integer> oldHouseholdSize, Map<Id, Contact> oldcmap, Contact c) {
+        return oldHouseholdSize.get(oldcmap.get(c.id).npo02__household__c) == 0;
+    }
+
+    private Id getOldHouseholdId(Map<Id, Contact> oldcmap, Contact c) {
+        return oldcmap.get(c.id).npo02__household__c;
```

Comment:
Doing a map.get() returns null if the key is not found, so there is a risk for a null pointer exception when trying to get the npo02__household__c from the map value.